### DISCUSSION TITLE
[Proposal] Support tail recursion

### DISCUSSION THREAD

**Discussion by gafter** [dotnet/csharplang#2544] [Language Ideas]
For performance reasons, it would be very helpful if some method invocations were to generate tail calls, in which the caller's stack is discarded and replaced by the callee's stack. When the method is calling itself, the compiler can generate a jump to the start of the method.

There are a few ways one might designate that one desires tail calls in the generated code:
1. The compiler could _always_ generate tail calls when applicable.
2. The compiler could generate tail calls when applicable in production builds but not debug builds. Equivalently, the compiler could recognize a command-line option (or project-wide option) that causes the generation of tail calls wherever possible.
3. The language could provide a syntax for tail invocations.
4. The compiler could recognize an attribute on the enclosing method and generate tail calls in the method's body wherever possible.

The advantages of a tail invocation is that it reduces the amount of stack used in a recursive call chain. That may improve performance due to improved memory locality. For some kinds of recursive code, it would allow the code to run with unbounded recursion; that makes it practical (without loss of performance) to replace loops with equivalent recursive code where the recursive code is more clear.

The disadvantages are that it makes debugging a bit harder, as stack frames with tail calls disappear from the call stack. Similarly they disappear from an exception stack trace. If the compiler does it automatically, this would be a behavioral change from previous releases, and would "break" any client that depends on the shape of the stack trace.

**Comment by svick** [comment]
AFAIK the CLR already supports tail recursion in some cases. What exactly are the cases where it doesn't and wouldn't it make sense to improve the CLR for those cases, instead of the compiler?

**Comment by MattGertz** [comment]
Yes, the CLR supports .tail.  The Roslyn compiler currently has no facility for inserting .tail at all, regardless of what the CLR currently does, however.  This is unlike, for example, F#, where tail recursion is leveraged.

**Comment by VSadov** [comment]
Generally JIT emits tail calls when it finds that profitable. In some sense it is a hardware-specific optimization. Same call may benefit from tailcalling on x64 (where stack frames are larger) and not so much on x86 (where register set is limited). Unsurprisingly the JIT is much more aggressive with tail calls on x64.
I think for perf-oriented tailcalls it may still be better to leave it to the JIT as to a compiler that is closer to the metal.

There is another set of situations where tailcalls are used for correctness/safety. This is the case where a long chain of methods, one deferring to another (recursively or not) is expected. This is common in F# and is also a pattern used by DLR callsite binders. (http://dlr.codeplex.com/SourceControl/latest#DLR_Main/Runtime/Microsoft.Scripting.Core/Compiler/LambdaCompiler.Expressions.cs      look at EmitMethodCall method)

These are the cases where OpCodes.Tailcall  IL prefix is used to _force_ taicalling. The prefix basically says "please tailcall this even if that might be not profitable since we expect a long chain of calls".

This is a part that is not expressible in C#. Unlike inlining, which can be forced via attributes, tailcalling cannot be currently forced. If one needs to write the code like emitted by EmitMethodCall, he cannot use C#.

In this light it might be interesting to consider options *3 or *4 because, while rare, there are indeed known cases when user has reasons to force tailcalling.

**Comment by gafter** [comment]
The runtime generally does not do this automatically as that would make it impossible for the runtime to produce stack traces.

**Comment by tmat** [comment]
Seems like the cleanest way of doing this would be to add MethodImplOptions.AggressiveTailCall (similar to AggressiveInlining). Then either the compiler would emit tailcall prefix or the jit would just understand the flag.

**Comment by gafter** [comment]
@tmat Yes, that was option (4).

**Comment by tmat** [comment]
@gafter Kind of. I was trying to point out 1) that we can use an existing attribute and 2) the JIT could do this, not the compiler.

**Comment by gafter** [comment]
@tmat Oh, I didn't realize a suitable attribute already existed. When you said "add" I thought you meant we'd add the attribute.

Does the jit already do this? I suspect not since it affects the stack trace, but perhaps the attribute is license to do so.

**Comment by leppie** [comment]
@tmat Just attributing a method is not sufficient. You could have 2 paths where one is tail callable and the other not. This is probably best left to the JIT to decide. Tail calls in most cases are not cheap. 

OTOH, if you could explicitly indicate a tail call, the C# compiler could perform tail call elimination (iow turn it into a while loop, like F#/IronScheme does) which is a very good optimization.

**Comment by tmat** [comment]
@ngafter I meant to add a new value to an existing enum that is used by an existing attribute.

**Comment by paulomorgado** [comment]
A way of explicitly requesting for tail calls could help on debugging. Maybe some `MethodImplOptions`.

That way it would be a developer requested action. One would be valid to expect that, if the developer requested, he will know it when debugging. And it will also be explicit for anyone else working on that code.

**Comment by VSadov** [comment]
@gafter    JIT definitely does tailcals when running optimized code and not debugging.

The following program works and terminates normally on x64

``` C#
    class Program
    {
        static void Main(string[] args)
        {
            System.Console.WriteLine(Ping(int.MaxValue, 0));
        }

        public static long Ping(int cnt, long val)
        {
            if (cnt-- == 0)
                return val;

            return Pong(cnt, val + cnt);
        }
        public static long Pong(int cnt, long val)
        {
            if (cnt-- == 0)
                return val;

            return Ping(cnt, val + cnt);
        }
    }
```

Make sure the code is not forced to run on 32bit via "32bit preferred" (where tailcalling is considered generally not profitable) and is not running under debugger (which suppresses JIT optimizations).

===== Output:
2305843005992468481
Press any key to continue . . .

===== the native codegen for the   Ping:

``` ASM
            if (cnt-- == 0)
00000000  mov         eax,ecx 
00000002  lea         ecx,[rax-1] 
00000005  test        eax,eax 
00000007  jne         000000000000000D 
                return val;
00000009  mov         rax,rdx 
0000000c  ret 
0000000d  movsxd      rax,ecx 
00000010  add         rdx,rax 
00000013  mov         rax,7FFDDD600088h // address of Pong
0000001d  jmp         rax               // <-- tailcalling Pong here
```

This does interfere with stack traces/debugging. But that is basically release/debug tradeoff. Besides , there are existing attributes to selectively suppress JIT optimizations if needed.

Note that as long as performance is concerned, JIT can do a better job than C# here by choosing the fastest calling convention conditional on the actual execution environment. 
It is only if we have extra whole-program knowledge that Ping/Pong are mutually recursive and want to ensure that this code works regardless whether under debugger or on x86, then we want a way to _force_ tailcalls

**Comment by Sebazzz** [comment]
Isn't the point here that the generation (or not) of the tail call should be deterministic, and not based on environment or JIT. The only way to guarantee tail call is if the compiler generates them, not if the JIT does (which cannot be predicated).

**Comment by sharwell** [comment]
@gafter I think the first thing you need to decide is whether you want C# developers to be able to use tail call elimination as a key factor in algorithm implementation. For this to happen, developers must be able to rely on tail call elimination, which means two things:
1. The compiler must insert the `tail.` prefix on these calls
2. The source code must be able to express a semantic requirement that `tail.` be inserted on a specific call

If you don't want to provide the above, then really what you have is a behind-the-scenes optimization like inlining that C# developers may benefit from but shouldn't assume will or will not happen during any particular execution of an application.

Also keep in mind stack traces during most production executions were not deterministic, so changing them isn't really a breaking change. They already changed in the past.

**Reply by madelson** [reply]
> The source code must be able to express a semantic requirement that tail. be inserted on a specific call

This is really important. Because of stack overflow, the presence of this optimization being applied or not can make the different between a program working vs. crashing. If I had some code that relied on tail calls to avoid stack overflow, I'd want a way to have the compiler assert that it is actually generating tail instructions so that if someone comes along later and breaks the tail call they'll know about it.

**Comment by AdamSpeight2008** [comment]
How does F# handle debug and tail recursion?

**Comment by orthoxerox** [comment]
One option would be to use something like `return from function(params)` to indicate you want a tail call. This way the compiler would be able to check if tail recursion is actually possible (e.g. `return from n*fact(n-1)` would be an error.

**Comment by gafter** [comment]
The jit generally ignores the tail instruction for virtual calls. This would therefore require CLR changes.

**Comment by leppie** [comment]
@gafter: Which JIT? I have not seen this behavior in IronScheme where tail calls are crucial. I have seen issues however with Mono's 64-bit JIT. Edit: It could be that I have not noticed the problem due to optimizations applied.

Update: Ran tests and cant replicate: https://gist.github.com/leppie/add00f54e6bc180ea7d1

**Comment by gafter** [comment]
@leppie Your test doesn't have any virtual calls.

**Comment by leppie** [comment]
@gafter: 
https://gist.github.com/leppie/add00f54e6bc180ea7d1#file-test_decompiled-il-L78
https://gist.github.com/leppie/add00f54e6bc180ea7d1#file-test_decompiled-il-L39

Those are `SimpleClosure` inherited from `Closure` inherited from `Callable`.

**Comment by sharwell** [comment]
> The jit generally ignores the tail instruction for virtual calls. This would therefore require CLR changes.

You could also spec the C# feature to work with what you know works:
- The tail call operation is guaranteed to not add a stack frame in the following cases provided that the caller is not executing code in a synchronized region (try/catch/finally, lock, using, etc.):
  - The caller is an instance method, and the target of the invocation is a non-virtual or sealed instance method of the current instance (the compiler could safely emit a tail.call instead of tail.callvirt for this case)
  - The target of the invocation is a static method located within the same assembly
- In other cases, the tail call instruction is suggestive - the runtime is preferred but not required to invoke the target without adding a stack frame

**Comment by leppie** [comment]
@sharwell: The sealed concrete type might explain why I get reliable tail calls in IronScheme for overriden methods. I am still not sure I could use `call` instead of `callvirt` for those. I will test both unsealing type and using `call` to see what the effects are.

**Comment by orthoxerox** [comment]
Hm, I've implemented a quick and dirty version of the TCO, and I don't get SOEs over virtual calls in release mode, unless the JIT is smart enough to devirtualize them:

``` c#
    abstract class Abs
    {
        public abstract int Bar(int i);
    }

    interface IFoo
    {
        int Baz(int i);
    }

    class Foo : Abs, IFoo
    {
        public override int Bar(int i)
        {
            if (i <= 0)
                return i;
            return from Bar(i - 1);
        }

        public int Baz(int i)
        {
            if (i <= 0)
                return i;
            return from Bar(i - 1);
        }
    }

    static class Program
    {
        public static int Foo(int i)
        {
            if (i<=0)
                return i;
            return from Foo(i-1);
        }

        public static void Main(string[] args)
        {
            Abs abs = new Foo();
            IFoo ifoo = new Foo();
            Func<int, int> foo = null;
            foo = (i) =>
            {
                if (i <= 0)
                    return i;
                return from foo(i - 1);
            };

            WriteLine(abs.Bar(10000000));
            WriteLine(ifoo.Baz(10000000));
            WriteLine(foo(10000000));
            WriteLine(Foo(10000000));
            ReadKey(true);
        }
    }
```

Now to find out why this fails in debug mode with System.InvalidProgramException...

**Comment by leppie** [comment]
While this is nice, I think it should be a compiler switch for the guarantee, but allow the JIT to do it's own thing. 

When the switch is active, the compiler will need do a bit more work (ie TCE) to not make every call (in a tail position) a tail call (as this is a lot more expensive), and revert to normal calls when recursion is not unbounded (this is very hard outside the compiled code, but could be done for well known code).

Another option is something like `unsafe` behavior.

**Comment by leppie** [comment]
@orthoxerox Run PEVerify :D Probably a short code that needs to be long now (been there, got the T shirt ;p)

**Comment by orthoxerox** [comment]
@leppie No, these turned out to be extra opcodes that are there to break on the closing brace (https://blogs.msdn.microsoft.com/zelmalki/2008/12/11/why-are-these-nop-instructions-in-my-debug-build/)

**Comment by aidapsibr** [comment]
This feels to me like a language feature level implementation. I don't want the JIT or compiler deciding to or not to use tail recursion when I wrote my code to expect that behavior. If tail-recursion was to be added, it should be a new `tail` keyword in C# along with compiler checks. Is this what is being discussed?

**Comment by orthoxerox** [comment]
@psibernetic If you don't want the JIT to decide, then it's either a JIT change request (it doesn't do tail calls in debug mode even when you remove the breakpoint nop), or you actually rewrite the recursion into a loop in the compiler, like F# does (it is limited to self-recursion or to explicit groups of mutually recursive functions).

**Comment by MkazemAkhgary** [comment]
@psibernetic  i think it doesn't require keyword or special syntax. because this is only an optimization on certain type of recursive methods that can be replaced by loops or jumps.

@gafter we are nearly in 2018 and this is not yet supported? by the way to me option 2 and 4 seems most desirable. 

option 4 has no critical drawbacks, option 5 is to introduce compiler option.

**Comment by jaykrell** [comment]
This is a surprisingly complex topic and many people (including me?) only understand some of it.
People often ignore one part and exaggerate another.

Let me give what I know.

Sometimes tailcall is a performance win-win.
It can save CPU.
  jmp is cheaper than call/ret
It can save stack.
 Touching less stack makes for better locality.


Sometimes tailcall is a performance loss, stack win.
 The CLR has a complex mechanism in which to pass more parameters
 to the callee than the caller recieved. I mean specifically
 more stack space for parameters. This is slow. But it conserves stack.
 It will only do this with the tail. prefix.


If the caller parameters are stack-larger than callee parameters,
it usually a pretty easy win-win transform.
There might be factors like parameter-position changing from managed
to integer/float, and generating precise StackMaps and such.


Now, there is another angle, that of algorithms that demand tailcall
elimination, for purposes of being able to process arbitrarily large
data with fixed/small stack. This is not about performance, but about
ability to run at all.


People spoke of loops.
One must consider that tailcalls can be self-tailcalls or tailcall-to-other.
Typically, "when it matters", tailcall-to-other is a mutual recursion of 2
or more functions, forming a loop.


For portability, across languages (C#, Java, C++), runtimes (Mono, CoreCLR),
compiler and runtime optimization flags, it is difficult to rely on opportunistic
tailcall elimination. I have not heard of anything but .NET even having
the tail. prefix and its "strong encouragement". But perhaps Java has it?
I understand Scheme mandates, but what are the Scheme implementations?
For example, anything translating to C++ will be at the mercy of that which
they do not control, unless they transform to loops.

Some C++ programmers as well, are trained to avoid recursion, in order to use
bounded machine stack. Perhaps perhaps that is the best answer for its portability
and performance.

But it'd still be nice to see C# attempt to address this area more formally,
to allow the idiom of self-recursion and mutual-recursion to be executable
in fixed amount of stack.


While F# might be a great language, one should not have to leave C# to gain this feature.
I kinda wish there was some delineation in the language of "fast" and "expensive" tailcalls.
But it isn't just the stacked parameter size.
But maybe this is too low level for a programmer to think about.
The runtimes also have "extra parameters" for generics and, at least Mono, interface calls,
and "generic shared with value type", which provide additional tailcall challenges on some
architectures, e.g. arm32.

**Comment by jaykrell** [comment]
There are 2 kinds of tailcalls in .NET. There are fast ones, strictly faster & conserve stack. And there are slow ones that are slow and conserve stack.

You will only get the slow ones with explicit tail. prefix.

C++ surely only will do the fast ones.

The slow ones are very complicated, involving helpers and stack unwind.
They handle arbitrary caller/callee signature mismatch. I didn’t think this was possible but it is.

The fast ones are very simple, for when caller & callee signature match or roughly match. This is what most people probably think of.

**Comment by orthoxerox** [comment]
@jaykrell Are the fast ones you're taking about the `jmp <method>` ones?

**Comment by YairHalberstadt** [comment]
There are two different things being discussed here. One is emitting `.tail` calls in the compiler, and the other is tail call elimination by the compiler. Which one was this issue intended to discuss, or were they both on the table?

Is it worth splitting them out into 2 separate issues?

**Comment by orthoxerox** [comment]
@YairHalberstadt `.tail` calls are probably the wrong way forward. They are ignored by the runtime in debug configuration, compress the call stack and make it impossible to "break on the closing brace".

I'd rather have compiler support for explicitly self-recursive methods that are rewritten into loops, a la F#.

**Comment by YairHalberstadt** [comment]
@Orthoxerox
I agree. But this discussion doesn't make that clear. Is it worth creating a separate issue to track tail call elimination in the compiler via an attribute?

**Comment by YairHalberstadt** [comment]
> Handling local tail calls is an uninteresting corner case that can be done by the front-end compiler and has no effect on Roslyn as a consequence so I don't think it is worth discussing here.

Roslyn includes two front end compilers. One for C# and one for VB

**Comment by samuel-vidal** [comment]
_following comment moved from: https://github.com/dotnet/csharplang/issues/2304#issuecomment-493586405__

I think it is **an excellent idea**

_With C# supporting a very confortable style of functional programming it becomes increasingly valuable to have a construct to help guaranteeing stack safety._

If we agree this is the purpose of that proposal, I think we should precise its semantic as follows:

The proposed 'tail' syntax should mean exactly that : the compiler is asked to check that 1) the call is tail and 2) guarantee that the call will be optimized (either by the Jit through emitting the necessary IL or otherwise).

In my opinion it would be preferable to make the IL .tail work than to make it a compiler code rewrite (this is way too narrow, and way more complex).

A tail call is one that is performed just before exiting a method. Either returning (nothing) directly if that call returns void or return the result of that call in case it returns something. No operation on the return value should be performed (not even implicit cast, although upcast should be fine).

(In particular, not all recursive call are tail-recursive, and not all tail-calls need to be part of recursion.)

An other key point is that for a call to be tail, both methods have to share the same return type (can be relaxed a bit with covariance).

Also, assuming the criteria is implemented in the compiler, the .tail IL should be emitted even in the absence of the tail constraint. 

proposed syntax :
`
tail qualifiedmethodname(parameters);	// (the syntax is not tied to the return statement).
`

examples

```c#
void method1()
{
	tail method1();		// tail call
	return;
	...
	method2();		// tail call
	return;
}

void method2()
{
	method1();		// tail call
	return;
	...
	method3();		// Not tail
	return;
}

int method3()
{
	method1();		// Not tail
	return 1;
	...
	return method3();		// tail call
	...
	return method4() + 1;		// Not tail
}

int method4()
{
	return tail method3();		// tail call
	...
	tail method3();		// compilation fails with an error (or a warning)
	return 1;
	...
	retrun x == j ? tail method3() : method4() + 1;	// only the first call is tail.
}
```

A few points:

The real difficulty here, I think, will be the interaction with ref/in/out arguments (let alone stack alloc etc) rather than the tail call being to the same method.
A point to have in mind is that a virtual method calling itself is actually performing another virtual dispatch (could be confusing for the user).

**Scope** : The compilation should fail or produce a warning if the condition for guaranteeing tail call optimization by the Jit are not met (or too hard to figure out e.g. in the case of in/out/ref args etc).

**Comment by The-Futurist** [comment]
Isn't the primary motive for tail recursion to avoid stack overflow? Recursion is the norm in functional languages but far less common in most imperative languages. So just how many people or scenarios would benefit from this?

This sound like a feature that would benefit just a tiny fraction of use cases where C# is used. The effort should be devoted to more pressing and immediate language enhancements that yield a bigger payoff.

**Comment by spydacarnage** [comment]
I would use recursion a lot more if tail recursion was a thing we could have

**Comment by gafter** [comment]
@Korporal Many Roslyn compiler failures can be traced to the fact that the compilers are recursive on the shape of the program.  Eliminating unnecessary frames would relax or eliminate many program size and shape constraints that cause the compiler to blow up.  So it would be at least very useful for the Roslyn compiler.  And every user of the Roslyn compiler, even if they don't use the feature directly.

**Comment by AartBluestoke** [comment]
The problem is that the following pattern is fundamentally dangerous unless you can GUARANTEE either:
a) quick termination of recursion (limited by an unknown stack size limitation)
b) --an-eventual-termination-- very deep recursion , AND tail recursion
edit: responding to the follow-up comment by  awshelley-beckman - the guarantee was concerning the picking of A or B, not the termination of the recursion, we do a) and explode with stack overflow at the moment, and b) tail recursion would allow potential infinite recursive functional programming patterns.

Because the first eliminates a large number of use cases (would you annotate a function "probably crashes with stack overflow if passed a list of more than 32k elements, or half that if you use this in ASP.Net")

This feature would be great. A DEMAND for tail recursion or compilation failure/warning? would allow us to actually use "potentially" tail recursive scenarios when you know it could loop 1 million times. 

At the moment even if roslyn reliably supported tail recursion, Using tail recursion would still be a "gosh, i hope this compile actually produces the specific byte code pattern that i *REQUIRE* for sane program behaviour".

When the creator of a program requires specific compile pattern for correctness, it would be helpful if the language/ compiler supports that request.

This results in recommendations to not use tail recursion unless your algorithm is O(log n) (eg https://softwareengineering.stackexchange.com/a/189537/111157  )

Code pattern many people would like to guarantee to be safe:
```
thing f(something){
..
return f(modifiedSomething);
}
```

**Reply by awshelley-beckman** [reply]
> b) an eventual termination, AND tail recursion

I would argue that this isn't necessary.  The language has no problem with `while (true)` loops that never terminate.  Why should tail call recursion be any different?

> At the moment even if roslyn reliably supported tail recursion, Using tail recursion would still be a "gosh, i hope this compile actually produces the specific byte code pattern that i REQUIRE for sane program behaviour".

Perhaps a special "tail" keyword could be added, which tells the compiler that you want this call to be a tail call.

**Reply by timcassell** [reply]
> Perhaps a special "tail" keyword could be added, which tells the compiler that you want this call to be a tail call.

I like that idea.

**Reply by agocke** [reply]
#2304

**Comment by weltkante** [comment]
Recently I'm repeatedly running into this issue when implementing custom awaitables and async primitives. With awaiters you don't have any choice about being iterative or recursive, the `OnCompleted` callbacks force you to be recursive. ~~I could really have used the ability to do a guaranteed tailcall to a delegate in the language.~~

~~Just throwing that in as a vote for "there is demand".~~

[edit] I was mistaken, tail recursion doesn't help my case, it just looked like it because I'm doing a call+return as last thing everywhere in the algorithms, but the delegates passed around can't be made tail recursive so you still get a stack of user code even if the rest could be optimized. Sorry for the noise.
